Add support for ADP5055

actorreno opened the pull request:
## PR Description

- This adds driver support and device tree bindings for ADP5055 triple buck regulator.
- Datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/adp5055.pdf
- Tested on RPI4 on the eval board: https://www.analog.com/media/en/technical-documentation/user-guides/adp5055-evalz-ug-1930.pdf

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented:
Please, fix first the valid CI errors you have...

actorreno commented:
It seems the checks returns some errors I couldn't capture when locally running commit check scripts like "./scripts/checkpatch <driver_file>"

I have fixed them except 1 last error which I am confused about since the error is not related to my files.
Screenshot below.

![image](https://github.com/user-attachments/assets/4117d006-bf3e-483f-b34f-52d1a6395d19)

kister-jimenez reviewed (commented):
ADP5055 is PMBUS compatible, should this be under hwmon/pmbus?

actorreno commented:
The datasheet did say "pmbus compatible" but it didn't have any pmbus commands the other typical pmbus dataseet contain
it also doesn't have an eeprom for the usual "pages" I see in other pmbus codes.

some colleagues agreed with me that it was just an i2c interface and decided it is better off in the regulator subsystem

nunojsa commented:
> ADP5055 is PMBUS compatible, should this be under hwmon/pmbus?

Hmm, I do think this is a pertinent question... Looking at the description:

"The ADP5055 is a power management unit that combines three high performance buck regulators with a PMBus interface..."

So this looks like a typical power management device that fits in HWMON/PMBUs. Note that the pmbus subsystem also has an interface with the regulator subsystem and it will also check what standard PMbus registers are supported by the device.  On top of that, you can pretty much bypass the standard registers and control the device as you want. Example, I see the VIDx_LOW/HIGH registers that could fit in the standard HWMON ABI and you totally support it even not using standard pmbus commands. There's also power good and overcurrent stuff that could very likely be made to work.

All the above to say that pmbus might be doable and allows you to better support this device (meaning, more features). You can also internally reach the apps engineer for this part  and try to understand the actual  PMBUS commands supported. Looking at the datasheet, I can see STATUS_CML to be supported.

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,203 @@
+      0 - Physical EN pin is used for enabling.
+      1 - Internal Register is used for enabling.
+      2 - Both register and physical pin needed.
+      3 - Either register or physical pin.
```
This looks custom implementation of a subset of the ON_OFF_CONFIG command. Moreover, DT maintainers do not like much of properties like this where you just give the register code. You should have a string property and do the mapping in the driver (for register values). On top of that, you should likely control 3 gpios that should be mandatory for options 0 and 3.

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,203 @@
+      Enables or disables over current protection
+      blanking (OCP) for all channels.
+      0 = Disabled
+      1 = Enabled
```
no need for the 0/1 description. Should encode the state in the property name. Simply `adi,ocp-blanking` and if the property is enabled, then it's pretty clear we're enabling the feature.

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,203 @@
+    items:
+      minimum: 0
+      maximum: 3
+    default: 3, 3, 3
```
You have a lot properties for the channels so I think it might be a good idea to create channel child nodes and have the properties per channel instead of this.

nunojsa reviewed (commented):
I'm still not convinced on the location of the driver but i'm already doing some comments on the DT bindings

actorreno commented:
> > ADP5055 is PMBUS compatible, should this be under hwmon/pmbus?
> 
> Hmm, I do think this is a pertinent question... Looking at the description:
> 
> "The ADP5055 is a power management unit that combines three high performance buck regulators with a PMBus interface..."
> 
> So this looks like a typical power management device that fits in HWMON/PMBUs. Note that the pmbus subsystem also has an interface with the regulator subsystem and it will also check what standard PMbus registers are supported by the device. On top of that, you can pretty much bypass the standard registers and control the device as you want. Example, I see the VIDx_LOW/HIGH registers that could fit in the standard HWMON ABI and you totally support it even not using standard pmbus commands. There's also power good and overcurrent stuff that could very likely be made to work.
> 
> All the above to say that pmbus might be doable and allows you to better support this device (meaning, more features). You can also internally reach the apps engineer for this part and try to understand the actual PMBUS commands supported. Looking at the datasheet, I can see STATUS_CML to be supported.

-----------------------------------------------------------------------------------------------------------------------------------------------
Hi,

I understand the reason on why pmbus should be reconsidered and I have been trying to learn how regulators are incorporated in the subsystem. However, it seems that when using "pmbus_do_probe" it calls a function called "pmbus_init_common" that looks for specific registers mainly PMBUS_STATUS_BYTE or PMBUS_STATUS_WORD (address 0x78 and 0x79) and this doesn't exist in the device hence the driver cannot properly load.

A manual i2c transaction does yield a read error
![image](https://github.com/user-attachments/assets/a2b5761b-66a7-4f17-969f-1ca08ad10a79)

Keeping it under the regulator subsystem still seems appropriate due to this.

Will yet to address dt-bindings feedbacks.

actorreno commented:
v2

yaml:

- adjusted to include child nodes for some properties. 
- added string properties where needed

.c:

- adjusted to read the new types of properties like from in to string, from array to child nodes

nunojsa commented:
> that looks for specific registers mainly PMBUS_STATUS_BYTE or PMBUS_STATUS_WORD

Yeah that looks to be the case yes... We would need to change the pmbus core to not check for these registers or a way to use driver specific stuff but yes, at this point I would not bother for that. Let's keep it under the regulator subsystem and hope no one asks for the monitoring features

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,227 @@
+        $ref: /schemas/types.yaml#/definitions/uint32
+        minimum: 0
+        maximum: 7
+        default: 0
```
The above two properties could also be an enum in milliseconds. Will produce an enum with 16 possible values which is a bit big but likely more user friendly than having the raw register values

actorreno replied:
Would it be better to keep this format? since changing the value of Tset requires soldering a different resistor to the CFG2 pin.

Though I agree I may need to edit the entry to be more user friendly so am thinking of calling the property a delay multiplier of Tset instead of "the delay itself"

nunojsa replied:
yes, but we do not have that many possibilities for TSET right? That is why I mentioned 16 entries... Anyways, not going to push for it but bear in mind that DT maintainers can complain about this

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,227 @@
+  - reg
+  - hw_en_0-gpios
+  - hw_en_1-gpios
+  - hw_en_2-gpios
```
the GPIOs are not required AFAICT...

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+#include <linux/regmap.h>
+#include <linux/gpio/consumer.h>
+#include <linux/of_gpio.h>
+#include <linux/of.h>
```
if you have this, then you need to depend on OF in your Kconfig

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+#include <linux/regulator/driver.h>
+#include <linux/regmap.h>
+#include <linux/gpio/consumer.h>
+#include <linux/of_gpio.h>
```
I dunno the above is needed

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+#include <linux/gpio/consumer.h>
+#include <linux/of_gpio.h>
+#include <linux/of.h>
+#include <linux/bitfield.h>
```
headers should come in alphabetical order

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+
+static const struct linear_range adp5055_voltage_ranges[] = {
+	REGULATOR_LINEAR_RANGE(408000, 0, 255, 1500),
+};
```
you need to select REGMAP_I2C in kconfig

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+		ret = of_property_read_u32(child, "adi,fast-transient-ch",
+					&adp5055->fast_transient_ch[i]);
+		if (ret)
+			return ret;
```
you're mixing OF properties with device. Please only use the device agnostic API...

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+		adp5055->enable_mode = ret;
+	adp5055->ocp_blanking = device_property_read_bool(dev, "adi,ocp-blanking");
+
+	for_each_child_of_node(np, child) {
```
[device_for_each_child_node_scoped()](https://elixir.bootlin.com/linux/v6.13.1/C/ident/device_for_each_child_node_scoped)

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+		adp5055->dvs_limit_lower_ch[i] = 0;
+		adp5055->fast_transient_ch[i] = 3;
+		adp5055->mask_power_good_ch[i] = 0;
+	}
```
Drop the all the 0 assignments. The memory is already initialized to 0 when allocated

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+		ARRAY_SIZE(adp5055_enable_mode_vals));
+	if (ret >= 0)
+		adp5055->enable_mode = ret;
+	adp5055->ocp_blanking = device_property_read_bool(dev, "adi,ocp-blanking");
```
As stated, for now I would drop the above property

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+	for_each_child_of_node(np, child) {
+		ret = of_property_read_u32(child, "reg", &i);
+		if (ret || i >= ADP5055_NUM_CH)
+			continue;
```
This is an error... treat it as such. `dev_err_probe()`

actorreno replied:
code snippet will be removed once 'for_each_child_of_node" is changed to device_for_each_child_node_scoped

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+
+		ret = of_property_read_u32(child, "adi,disable-delay-ch",
+					   &adp5055->disable_delay_ch[i]);
+		if (ret < 0) {
```
I guess you did not tested this... you mean `if (!ret)`

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+			ret = adp5055_get_prop_index(adp5055_disable_delay_ch_vals,
+						     ARRAY_SIZE(adp5055_disable_delay_ch_vals),
+						     adp5055->disable_delay_ch[i]);
+			if (ret < 0)
```
please add some logs for helping to understand the problem... `dev_err_probe()`

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+					adp5055_enable_mode_vals,
+					ARRAY_SIZE(adp5055_enable_mode_vals));
+		if (ret >= 0)
+			adp5055->enable_mode = ret;
```
again this property?

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+
+	val1 = FIELD_PREP(ADP5055_MASK_OCP_BLANKING, adp5055->ocp_blanking);
+	val2 = FIELD_PREP(ADP5055_MASK_PSM321, adp5055->power_saving_mode_ch[0]);
+	val3 = FIELD_PREP(ADP5055_MASK_DIS, adp5055->output_discharge_function_ch[1]);
```
/If this variables are only used in here you can just have some local variables instead of keeping them in your state struct. Also no need for val1, val2 and val3. Why not just?

`val1 |= FIELD_PREP(ADP5055_MASK_PSM321, adp5055->power_saving_mode_ch[0]);` and so on...

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+	val2 = FIELD_PREP(ADP5055_MASK_EN_DLY, adp5055->enable_delay_ch[2]);
+	ret = regmap_write(regmap, ADP5055_DLY2, val1 | val2);
+	if (ret)
+		return ret;
```
YOu can have this much better. Create a proper adp5505_chan_setup() function that you call in the end of your [device_for_each_child_node_scoped()](https://elixir.bootlin.com/linux/v6.13.1/C/ident/device_for_each_child_node_scoped) iteration. Then, no need for the above repetition and another plus is that you only configure channels that really need to be configured. i.e, no need for rewriting defaults.

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+		return 0;
+	} else {
+		return -EINVAL;
+	}
```
redundant else if() and else

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+
+static struct i2c_driver adp5055_driver = {
+	.driver	= {
+		.name	= "adp5055",
```
please add an of_id_table

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+	.probe		= adp5055_probe,
+	.id_table	= adp5055_ids,
+};
+
```
nit: no need for new line

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+		}
+	}
+
+	i2c_set_clientdata(client, adp5055);
```
is the above needed? does not looks like it is

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+
+		rdev = devm_regulator_register(dev, desc, &config);
+		if (IS_ERR(rdev)) {
+			dev_err(dev, "Failed to register %s\n", desc->name);
```
`dev_err_probe()`

nunojsa replied:
Yes, but you can use PTR_ERR() in dev_err_probe(). There are also other variants of dev_err_probe()

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,511 @@
+	if (!adp5055)
+		return -ENOMEM;
+
+	adp5055->regmap = devm_regmap_init_i2c(client, regmap_config);
```
why not passing directly adp5055_regmap_config?

actorreno commented:
(still correcting CI check)

v3:

.yaml:

-removed copyright
-removed description on "properties -> reg"
-added vendor prefix to gpios
- regarding enable-mode property, I initailly coded the options but
reduced this now to boolean. if 1, SW enable, if none use HW GPIOs
- removed 0,1 definition for delay-power-good, also stated Tset value as suggested
- removed mention of True False for boolean 'mask-power-good', plus uin32 ref removed
- for dvs upper/lower limit, changed from 4 bits code to microvolts. 
- Calling enable/disbale-delay-ch as multiplier rather than a delay
- removed GPIOs from "required"

.c:

- removed OF header (not needed anymore) to not change kconfig
- removed of_gpio.h
- alphabetized headers
- removed OF properties, now uses device_ or fwnode_
- dropped all initialization to 0
- edited enable-mode to a boolean to choose if sw or hw enable. if this
is still too much, can remove everything including GPIOs next version
- added some dev_err_probe and logs
- removed duplicate property
- kept previous state variables to just local ones
- compacted val1-val3 as just val
- regarding repetitive initiation using regmap_write, placed some inside device_for_each_child_node_scoped
- removed redundant if-else
- added an o_match_table/of_device_id
- removed newline before module_device_table
- removed i2c_set_clientdata
- kept dev_err, still curious why dev_err_probe should replace it unlike other regulator driver
- passed &adp5055_regmap_config instead

Kconfig:

- select regmap_i2c

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,216 @@
+    description:
+      If present, configure channels to be enabled via software
+      instead of hardware enable.
+    type: boolean
```
I do not think we need this. Can't we assume that if the gpios are not given, then we have SW enabled mode?

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,216 @@
+          Vout_low = Vout - DVS_lower_limit.
+        $ref: /schemas/types.yaml#/definitions/uint32
+        minimum: -190500
+        maximum: -10500
```
are the above in uV? If so you should add the proper suffix in the property and drop $ref. See the available standard suffixes here:

https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/property-units.yaml#L81

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+#include <linux/device.h>
+#include <linux/gpio/consumer.h>
+#include <linux/i2c.h>
+#include <linux/module.h>
```
You should include mod_devicetable.h for of_device_id

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+	if (IS_ERR(adp5055->regmap)) {
+		ret = PTR_ERR(adp5055->regmap);
+		dev_err(dev, "Failed to allocate register map: %d\n", ret);
+		return ret;
```
This can simply be:

`return dev_err_probe(dev, PTR_ERR(adp5055->regmap), "...");`

Ditto for all other error logs during probe()

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+		.enable_reg		= ADP5055_CTRL123,		\
+		.enable_mask		= BIT(_ch),			\
+		.owner			= THIS_MODULE,			\
+	}
```
I would drop the tabs and just use spaces...

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+	.enable				= adp5055_enable,
+	.disable			= adp5055_disable,
+	.is_enabled			= adp5055_is_enabled,
+};
```
ditto

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+	int mask;
+
+	id = rdev_get_id(dev);
+	mask = 1 << id;
```
BIT(id)

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+	if (adp5055->enable_mode_sw)
+		return (val_sw & mask) != 0;
+	else
+		return val_hw;
```
redundant else

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+
+	id = rdev_get_id(dev);
+	mask = 1 << id;
+	ret = regmap_read(adp5055->regmap, ADP5055_CTRL_MODE1, &val_sw);
```
check the error code

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+	int i, ret;
+	struct regmap *regmap = adp5055->regmap;
+	int val;
+
```
no need for this new line

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+							ARRAY_SIZE(adp5055_disable_delay_ch_vals),
+							enable_delay_ch[i]);
+			if (ret < 0)
+				return dev_err_probe(dev, ret, "Failed to initialize enable-delay-ch. Using default.");
```
80 column limit... You should break after ret I guess

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,434 @@
+	if (IS_ERR(adp5055->hw_en_gpio[2])) {
+		return dev_err_probe(dev, PTR_ERR(adp5055->hw_en_gpio[2]),
+				     "Failed to get hw_en_2 GPIO\n");
+	}
```
no need for {} when we just have one liner

actorreno commented:
v4

yaml:
- turned 3 gpios into a gpio array
- turned the delay from multipleir to actual delay in ms
-added a tset property to change enum of dis/enable-delay-ch
-edited dvs-limit-upper/lower default value actual value, it was 0 before
-added suffix microvolt to the property and removed $ref
-edited examples to match entry of yaml property list

c:
- edited code for changes in yaml
-added mod_devicetable.h
-simplified error probes in probe()
- dropped tabs and just used a space
- removed redundant 'else'
- added error code check in one regmap_read that had none
- removed newline in between variable declarations
- fixed 80 column limit issue
- removed {} for one liner if/else
- needed to add dvs-limit-upper/lower volts to register value code
- placed gpio array check earlier than other properties

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,212 @@
+      specific resistor values on the CFG2 pin. If property is absent,
+      the setting time Tset defaults to 2.6ms. If present, Tset is
+      increased 8 times to 20.8ms.
+    type: boolean
```
Why not having this with default suffixes? Just use an enum with two possible values with a default. I guess TSET has a default value if no resistor is soldered?

actorreno replied:
actually the fact the supposed enum is only two choices made me think of using bool. will use an enum similar to the microseconds feedback below

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,212 @@
+      hardware enable for channels 0-2. Should be marked 0 for active
+      low. Requires all three channels to be initialized. Not adding
+      the property turns the system to a software only enable mode.
+    maxItems: 3
```
I think you also need minItems. IIRC, this is an all or nothing property. So we need to require all three GPIOs

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,212 @@
+        description:
+          If present, enables power saving mode for
+          individual channels.
+        type: boolean
```
I would remove the -ch for all properties. It's already clear these properties are for the channel

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,212 @@
+        description:
+          Configures the disable delay for each channel. Dependent on Tset.
+        enum: [0ms, 5.2ms, 10.4ms, 15.6ms, 20.8ms, 26ms, 31.2ms, 36.4ms]
+        default: 0ms
```
Oh, this is not how it should be done... Moreover you'll need the property to be in microseconds. This should be:

```
adi,disable-delay-us:
  description:
    Configures the disable delay for each channel. Dependent on Tset.
  enum: [0, 5200, 10400, 15600, 20800, 26000, 31200, 36400]
  default: 0
```

Ditto for all the other properties

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,adp5055-regulator.yaml:
```diff
@@ -0,0 +1,212 @@
+  - if:
+      properties:
+        adi,tset-8x: true
+
```
nit: no need for extra line

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,438 @@
+	"104ms",
+	"124.8ms",
+	"145.6ms",
+};
```
you'll need to refactor the above based on latest feedback

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,438 @@
+	device_for_each_child_node_scoped(dev, child) {
+		ret = fwnode_property_read_u32(child, "reg", &i);
+		if (ret || i >= ADP5055_NUM_CH)
+			return dev_err_probe(dev, ret, "Failed to read reg value of child node");
```
please separate the checks. It's checking for different things and the error message will be misleading in case i >= ADP5055_NUM_CH

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,438 @@
+					&dvs_limit_lower_ch[i]);
+		if (ret)
+			return ret;
+
```
The above two do not exist in the bindings and are not mandatory...

nunojsa replied:
The above was not addressed...

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,438 @@
+	id = rdev_get_id(dev);
+	mask = BIT(id);
+
+	if (adp5055->enable_mode_sw) {
```
Where are you setting the above flag?. Anyways, I think you can use the hw_en_array_gpios pointer:

```
if (!adp5055->hw_en_array_gpios)
    return regmap_update_bits(adp5055->regmap, ADP5055_CTRL_MODE1, mask, en_val);

gpiod_set_value_cansleep(adp5055->hw_en_array_gpios->desc[id], en_val);
return 0;
```

actorreno commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,438 @@
+
+	gpiod_set_value_cansleep(adp5055->hw_en_array_gpios->desc[0], 0);
+	gpiod_set_value_cansleep(adp5055->hw_en_array_gpios->desc[1], 0);
+	gpiod_set_value_cansleep(adp5055->hw_en_array_gpios->desc[2], 0);
```
i see, all this time I thought when requesting the pins the GPIOD_OUT_LOW/HIGH just tells the "active" level of a pin.
then a gpiod_set..... will actually initialize it to a value we want. will remove

actorreno commented:
v5

yaml:

- changed tset-8x to tset-us and the enum choices
- added minItems for gpio property
- removed -ch for all channel properties
- edited dis/enable-delay-us with actual values in microsec not string

c:

- edited code to match yaml feedback on ms values
- turned devm_gpiod.. to devm_gpiod_get_array_optional()
- for the check if (adp5055->hw_en_array_gpios->ndescs != ADP5055_NUM_CH), used
the !adp5055->hw_en_array_gpios
- separated checks under fwnode_property_read_u32
- edited the two properties to match the missed new name in bindings
- removing all "adp5055->enable_mode_sw" populated by device_property_present
and replacing it with checks using "!adp5055->hw_en_array_gpios"
- removed gpiod_set_value_cansleep that just initializes the gpio pin. simplified to GPIOD_OUT_LOW

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,486 @@
+	if (!adp5055->hw_en_array_gpios) {
+		ret = regmap_update_bits(adp5055->regmap, ADP5055_CTRL_MODE1, mask, en_val);
+		if (ret)
+			return ret;
```
This can be simplified... Do `return regmap_update_bits()` and then you do not need the else {

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,486 @@
+	if (ret)
+		return ret;
+
+	val_hw = gpiod_get_value_cansleep(adp5055->hw_en_array_gpios->desc[id]);
```
move this after the if(). No need to do it if not in hw mode

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,486 @@
+
+	adp5055->hw_en_array_gpios = devm_gpiod_get_array_optional(dev,
+				"adi,hw-en-array", GPIOD_OUT_LOW);
+
```
nit: no new line

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,486 @@
+		return dev_err_probe(dev, PTR_ERR(adp5055->hw_en_array_gpios),
+				     "Failed to get hw_en_array GPIOs\n");
+
+	if (!adp5055->hw_en_array_gpios)
```
I think you mean `if (adp5055->hw_en_array_gpios)`

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,486 @@
+				     "Failed to get hw_en_array GPIOs\n");
+
+	if (!adp5055->hw_en_array_gpios)
+		if (adp5055->hw_en_array_gpios->ndescs != ADP5055_NUM_CH)
```
a dev_err_probe() log message could be useful...

nunojsa commented on drivers/regulator/adp5055-regulator.c:
```diff
@@ -0,0 +1,486 @@
+	ret = device_property_read_u32(dev, "adi,tset-us", &tset);
+	if (!ret) {
+		ret = adp5055_get_prop_index(adp5055_tset_vals,
+						ARRAY_SIZE(adp5055_tset_vals), tset);
```
nit: align on open parenthesis...

nunojsa reviewed (commented):
Feel free to take my last comments and send this upstream... I would also recommend to improve the commit message for the bindings:

1. Remove the .yaml suffix on the commit subject;
2. Add a meaningful message to the commit message. Saying "Adding bindings docs" says nothing and the maintainers will likely complain. Give a small description of the device.

actorreno commented:
This version shall be sent upstream

v6

yaml:

- dropped tset-us description "Soldering nothing results to the default value in microseconds."
- edited commit subject and added more on the message to describe device

c:

- simplified an if else with {} by 'return regmap_update_bits()'
- moved a code used during HW_en after the "hw_en check" to lessen code during sw_en
- removed new line after 'adp5055->hw_en_array_gpios = devm_gpiod_get_array_optional', error check now directly follows
- changed an if by removing '!' -> if (adp5055->hw_en_array_gpios) during a check where the code needs to make sure it exist first
- added dev_err_probe about not having the valid amount of channels
- adjusted indent to align on open parenthesis
- removed the for loop initializing fast_transient due to a check, putting the default in case of error
- dvs-limit-lower/upper now not mandatory code wise, will default in case of error. also added Out of range check

Note:

the CI check for the dtbinding for this version had a package error and is not related to the code.  the changes done for this version is only editing the description so should be no error from previous. see image below of azure error msg
![image](https://github.com/user-attachments/assets/ef690219-191b-4166-8159-c74e457c38da)

actorreno commented:
v7: (Series Applied Upstream)

- Multiple properties removed and moved to core regulator ops
- dt property parsing moved to of_parse_cb

Applied to:


[https://urldefense.com/v3/__https://git.kernel.org/pub/scm/linux/kernel/git/broonie/regulator.git__;!!A3Ni8CS0y2Y!9NCSufd1RqOYSqg15_jiLcV7MTZdLrevrlyimxmhba6TmW1rSi4697JYXeKtPv6gGsYZchPftZea3tOBGPB6JUFusQ$](https://urldefense.com/v3/__https:/git.kernel.org/pub/scm/linux/kernel/git/broonie/regulator.git__;!!A3Ni8CS0y2Y!9NCSufd1RqOYSqg15_jiLcV7MTZdLrevrlyimxmhba6TmW1rSi4697JYXeKtPv6gGsYZchPftZea3tOBGPB6JUFusQ$)  for-next

Thanks!

[1/2] regulator: dt-bindings: adi,adp5055-regulator: Add adp5055 support
      commit: 5bfc88c4e5783357399ce66f40a479a090af2ba0
[2/2] regulator: adp5055: Add driver for adp5055
      commit: 147b2a96f24e0cfcc476378f9356b30662045c7e

-----------------------------------------------------------------------------------

CI issue:

Check patch fails here while not during upstream check (b4 prep --check). See image below:

![image](https://github.com/user-attachments/assets/c5658879-4b31-4609-8b3f-e2211751a336)

Error is re-use of arguments. 
It seems inconsequential but am not sure how to go about this. Can this be accepted as is or must work around the error?
